### PR TITLE
chore: add custom node taints

### LINKED ISSUE #7581: Support for Specifying Custom Node Taints via Machine Configuration
**Description**:

In the current implementation of Talos, there is no native support for specifying custom node taints directly through the machine configuration. While control plane nodes are automatically tainted, other nodes lack this configurability.

**Problem Statement:**

Node taints are a critical part of Kubernetes cluster management, enabling operators to control which nodes specific pods can be scheduled on. The lack of support for custom taints within Talos's machine configuration makes it difficult to manage taints at scale and can lead to inefficiencies and inconsistencies across the cluster.

**Current Workaround:**

At present, custom taints can be applied using the kubectl command after the node is installed. While this method does provide a way to apply taints, it introduces extra manual steps and can lead to inconsistencies in the configuration of nodes within a cluster. Automating this process within Talos would greatly streamline cluster configuration, ensuring that taints are applied uniformly and without requiring individual manual intervention.

**Proposed Solution:**

Introduce native support in Talos for specifying custom node taints within the machine configuration. This would allow taints to be defined as part of the machine's initial setup, ensuring a consistent configuration across all nodes in the cluster. The solution would provide cluster administrators with a more efficient and robust way to manage node taints, aligning with common Kubernetes best practices.

### PR DESCRIPTION
**Pull request by DmitriyMV**
This PR adds support for custom node taints. Refer to `nodeTaints` in the `configuration` for more information.

Closes #7581

### CODE REVIEW

File: pkg/machinery/labels/validate.go [hunk 1]
```diff
@@ -130,3 +130,28 @@ func ValidateLabelValue(value string) error {
 
 	return nil
 }
+
+// ValidateTaints validates that a set of taints is correctly defined.
+func ValidateTaints(taints map[string]string) error {
```
Comment (DmitriyMV) [inline pkg/machinery/labels/validate.go:135]: Is this correct way of validating taints?

**Comment by DmitriyMV** [comment]
There is also https://github.com/siderolabs/talos/blob/dbaf5c69978fd1d22737385ddd096798d408254c/internal/integration/api/node-labels.go#L85-L90 integration test. Should I update it to handle the taints too?

File: pkg/machinery/labels/validate.go [hunk 2]
```diff
@@ -130,3 +130,28 @@ func ValidateLabelValue(value string) error {
...
+			multiErr = multierror.Append(multiErr, fmt.Errorf("invalid taint effect: %s", effect))
+		}
+
+		if err := ValidateLabelValue(val); err != nil {
+			multiErr = multierror.Append(multiErr, err)
+		}
+	}
+
+	return multiErr.ErrorOrNil()
+}
```
Comment (DmitriyMV) [inline pkg/machinery/labels/validate.go:157]: Is this correct way of validating taints?

**Comment by smira** [comment]
> integration test. Should I update it to handle the taints too?

yes, absolutely, probably another test which will be very similar

File: pkg/machinery/labels/validate.go (hunk 2 above)
Comment (smira) [reply pkg/machinery/labels/validate.go:157]: you can check with either k8s docs or their code (we can't use it in machinery, but you can get the idea there)

**Review by smira — COMMENTED** [review]
lgtm, just need the integration test!

File: pkg/machinery/labels/validate.go [hunk 3]
```diff
@@ -130,3 +130,28 @@ func ValidateLabelValue(value string) error {
...
+	keys := maps.Keys(taints)
+	slices.Sort(keys)
+
+	for _, k := range keys {
+		if err := ValidateQualifiedName(k); err != nil {
+			multiErr = multierror.Append(multiErr, err)
+		}
+
+		val, effect, found := strings.Cut(taints[k], ":")
+		if found && effect != "NoSchedule" && effect != "PreferNoSchedule" && effect != "NoExecute" {
```
Comment (frezbo) [inline pkg/machinery/labels/validate.go:147]: doesn't k8s supports taints other than this :thinking: ?

Comment (frezbo) [reply pkg/machinery/labels/validate.go:147]: cool, maybe we keep this in `constants`? and then filter on a slice?

**Review by frezbo — APPROVED** [review]

File: pkg/machinery/labels/validate.go (hunk 3 above)
Comment (DmitriyMV) [reply pkg/machinery/labels/validate.go:147]: I dunno, both works for me.

Comment (DmitriyMV) [reply pkg/machinery/labels/validate.go]: Updated the code.

File: pkg/machinery/labels/validate.go (hunk 2 above)
Comment (DmitriyMV) [reply pkg/machinery/labels/validate.go]: Yup, it uses the same logic https://github.com/kubernetes/kubectl/blob/b35935138f5330d299e89b1278b5738487e4f015/pkg/cmd/taint/utils.go#L39

**Review by frezbo — APPROVED** [review]

**Pull request merged**